src/main/scala/fi/oph/koski/sso/CasServlet.scala

```diff
@@ -1,26 +1,72 @@
 package fi.oph.koski.sso
 
+import java.net.URLEncoder.encode
+import java.nio.charset.StandardCharsets
 import fi.oph.koski.config.KoskiApplication
 import fi.oph.koski.http.{Http, KoskiErrorCategory, OpintopolkuCallerId}
-import fi.oph.koski.koskiuser.{AuthenticationSupport, DirectoryClientLogin, KoskiUserLanguage}
+import fi.oph.koski.koskiuser.{AuthenticationSupport, AuthenticationUser, DirectoryClientLogin, KoskiUserLanguage}
 import fi.oph.koski.log.LogUserContext
 import fi.oph.koski.servlet.{NoCache, VirkailijaHtmlServlet}
-import fi.vm.sade.utils.cas.CasClient.Username
-import fi.vm.sade.utils.cas.{CasClient, CasLogout}
+import cas.{CasClient, CasClientException}
+import cas.CasClient.{OppijaAttributes, Username}
+import fi.oph.koski.henkilo.OppijaHenkilö
+import fi.oph.koski.json.JsonSerializer.writeWithRoot
+import scalaz.concurrent.Task
+import fi.oph.koski.schema.{Nimitiedot, UusiHenkilö}
+import org.scalatra.{Cookie, CookieOptions}
+
+import scala.util.control.NonFatal
 
 /**
   *  This is where the user lands after a CAS login / logout
   */
-class CasServlet(implicit val application: KoskiApplication) extends VirkailijaHtmlServlet with AuthenticationSupport with NoCache {
-  private val casClient = new CasClient(application.config.getString("opintopolku.virkailija.url"), Http.newClient("cas.serviceticketvalidation"), OpintopolkuCallerId.koski)
+class CasServlet()(implicit val application: KoskiApplication) extends VirkailijaHtmlServlet with AuthenticationSupport with NoCache {
+  private val casVirkailijaClient = new CasClient(application.config.getString("opintopolku.virkailija.url") + "/cas", Http.newClient("cas.serviceticketvalidation"), OpintopolkuCallerId.koski)
+  private val casOppijaClient = new CasClient(application.config.getString("opintopolku.oppija.url") + "/cas-oppija", Http.newClient("cas.serviceticketvalidation"), OpintopolkuCallerId.koski)
   private val koskiSessions = application.koskiSessionRepository
 
+  protected def onSuccess: String = params.get("onSuccess").getOrElse("/omattiedot")
+  protected def onFailure: String = params.get("onFailure").getOrElse("/virhesivu")
+  protected def onUserNotFound: String = params.get("onUserNotFound").getOrElse("/eisuorituksia")
+
   // Return url for cas login
-  get("/") {
+  get("/oppija") {
```

Comment:
Kuten sanoit, voisi tätä ja muita tämän servletin funkkareiden sisältöjä siistiä. Mun mielestä olisi ihan bueno joku CasService tms. luokka, johon siirtäisi logiikan ja servletin tehtäväksi jää pääasiassa kaivaa requestista parametrit ja passata ne eteenpäin.